Update to latest Snowdrop Buildpack Platform Impl

BarDweller opened the pull request:
Significant update to the buildpack platform implementation used by the buildpack container image extension. 

Updates from snowdrop lib 0.0.6 to 0.0.10 bringing support for builder images using image extensions, and updating the default builder to be the paketo ubi builder image. Also contains updates for podman compatibility, image pull retry, and various other minor fixes. 

- Fixes: #41899 
- Fixes: #37777 
- Fixes: #23883 
- Fixes: #23559

gsmet commented on extensions/container-image/container-image-buildpack/deployment/src/main/java/io/quarkus/container/image/buildpack/deployment/BuildpackConfig.java:
```diff
@@ -14,13 +14,13 @@ public class BuildpackConfig {
      * The buildpacks builder image to use when building the project in jvm mode.
      */
-    @ConfigItem
+    @ConfigItem(defaultValue = "paketocommunity/builder-ubi-base")
     public Optional<String> jvmBuilderImage;
```
Given you provided a default value, you should drop the `Optional`.

Same for the one just below.

It's not mandatory so I suppose we can live with it if you don't have to push other changes but if CI doesn't pass and you do, please drop the Optional. Thanks!

gsmet reviewed (commented):
Thanks! I added a small minor comment, either for this time or next depending on how the CI goes.

geoand commented:
What's the status of this?

BarDweller commented:
> What's the status of this?

I'm removing the optional, and also performing a few more tests locally with native builds, that have thrown a curveball.. I'll add the PR for the Optional & removal of the native builder default.. There will probably be a new PR for the native changes

BarDweller commented:
Summary of findings from post-review comments... 

 - builder default removed for native build, the paketo ubi builder does not do native builds, there is currently no available builder to configure for quarkus native builds.
 - testing performed locally to validate that native builds were possible for quarkus using a custom builder image derived from the quarkus mandrel container. 
 - testing uncovered issue in snowdrop lib related to builder images with unexpected entrypoint configurations. Snowdrop lib fixed, and re-released as 0.0.11
 - exposed a few more config options from the snowdrop lib via quarkus that were helpful during native testing. 
 
 Quarkus native builder image can be constructed relatively simply, although it ends up as a builder per java version, need to look into if the github project that publishes the quarkus mandrel images could also do buildpack builder images.. 
 
 Quarkus buildpack also needs an update, as it's using deprecated property names.. raised https://github.com/paketo-buildpacks/quarkus/issues/83 and will help get that updated.

geoand reviewed (approved):
LGTM

geoand commented:
@iocanel mind taking a look as well?

iocanel reviewed (approved):
LGTM

michalvavrik commented on extensions/container-image/container-image-buildpack/deployment/src/main/java/io/quarkus/container/image/buildpack/deployment/BuildpackConfig.java:
```diff
@@ -14,15 +14,21 @@ public class BuildpackConfig {
 
     /**
-     * The buildpacks builder image to use when building the project in jvm mode.
+     * The buildpacks builder image to use when building the project in native mode.pu
```
typo `mode.pu` ?

geoand commented:
Please squash the commits

iocanel commented on extensions/container-image/container-image-buildpack/deployment/src/main/java/io/quarkus/container/image/buildpack/deployment/BuildpackProcessor.java:
```diff
@@ -181,33 +182,48 @@ private String runBuildpackBuild(BuildpackConfig buildpackConfig,
                         if (buildpackConfig.dockerHost.isPresent()) {
                             log.info("Using DockerHost of " + buildpackConfig.dockerHost.get());
-                            b.withDockerHost(buildpackConfig.dockerHost.get());
+                            b.editDockerConfig().withDockerHost(buildpackConfig.dockerHost.get());
```
```suggestion
                            b.editDockerConfig().withDockerHost(buildpackConfig.dockerHost.get()).endDockerConfig();
```

iocanel commented on extensions/container-image/container-image-buildpack/deployment/src/main/java/io/quarkus/container/image/buildpack/deployment/BuildpackProcessor.java:
```diff
@@ -181,33 +182,48 @@ private String runBuildpackBuild(BuildpackConfig buildpackConfig,
-                    }).build();
+                        if (buildpackConfig.trustBuilderImage.isPresent()) {
+                            log.info("Setting trusted image to " + buildpackConfig.trustBuilderImage.get());
+                            b.editPlatformConfig().withTrustBuilder(buildpackConfig.trustBuilderImage.get());
```
```suggestion
                            b.editPlatformConfig().withTrustBuilder(buildpackConfig.trustBuilderImage.get()).endPlatformConfig();
```

iocanel reviewed (changes requested):
There is a minor issue with the the use of `editXXX` as its seems that the `endXXX` is missing. The intention is that the actual edited value is set when you actually call the `endXXX` method.

For example:

```java
   ObjectMetaBuilder builder;

    public N endMetadata() {
      return (N) MyFluent.this.withMetadata(builder.build());
    }
```

BarDweller commented:
Updating to add endXXXConfig and rerunning local tests.. will squash merge when done

BarDweller commented:
Updated, found issue with snowdrop library not honoring docker pull timeout/timeout increase amounts, fixed library, updated quarkus usage.. tested as functional.

cmoulliard commented:
What is the status ? @BarDweller

BarDweller commented:
Just needs a small git tidy up so it can be re-reviewed/merged.. (merge commit snuck in when I updated to latest branch.. need to rebase & force push to sort it out.. will be done by eod)

BarDweller committed:
Update to latest Snowdrop Buildpack platform impl

iocanel reviewed (approved):
LGTM
